### PR TITLE
Allow Share Dialogs to create Shortcuts

### LINKED ISSUE #3175: Add home screen folder/file shortcut widget
Requesting the ability to drop a shortcut on Android home screens that link to specific folders or files. Thanks!

### PR DESCRIPTION
**Pull request by newhinton**
- [ ] Tests written, or not not needed

Closes #3175 

This is based on #5522. 
Thanks to @wiwie!

This Pull request allows the user to create homescreen-shortcuts via the share dialog.

| Share for Folder         | Share for Files     |
|--------------|-----------|
| grafik | grafik |

### CODE REVIEW

File: app/src/main/java/com/owncloud/android/utils/ShortcutUtil.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
+package com.owncloud.android.utils;
```
Comment (AlvaroBrey) [inline app/src/main/java/com/owncloud/android/utils/ShortcutUtil.java:1]: We don't allow new Java files; this must be converted to Kotlin.

Additionally this file is missing a license header.

File: app/src/main/res/drawable/add_to_home_screen.xml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+<vector android:autoMirrored="true" android:height="24dp"
```
Comment (AlvaroBrey) [inline app/src/main/res/drawable/add_to_home_screen.xml:1]: missing license header

File: app/src/main/res/layout/file_details_share_pin_home.xml [hunk 3]
```diff
@@ -0,0 +1,69 @@
+<?xml version="1.0" encoding="utf-8"?><!--
+  ~
+  ~ Nextcloud Android client application
+  ~
+  ~ @author Tobias Kaminsky
+  ~ Copyright (C) 2020 Tobias Kaminsky
+  ~ Copyright (C) 2020 Nextcloud GmbH
```
Comment (AlvaroBrey) [inline app/src/main/res/layout/file_details_share_pin_home.xml:7]: wrong author and year

File: gradle.properties [hunk 4]
```diff
@@ -1,7 +1,19 @@
-# can be overriden by ~/.gradle/gradle.properties
-NC_TEST_SERVER_BASEURL=http://server
-NC_TEST_SERVER_USERNAME=test
+## For more details on how to configure your build environment visit
```
Comment (AlvaroBrey) [inline gradle.properties:1]: unrelated to this PR, please undo changes to this file

**Review by AlvaroBrey — CHANGES_REQUESTED** [review]
Hello @newhinton and thanks for your contribution! Here's a quick initial PR; once you fix these I will do a more in-depth one.

In addition to the below comment, GitHub also reports this PR has a conflict, please rebase on master and fix it.

**Comment by newhinton** [comment]
You are right, i will fix those issues! 

However, have the design/ui/ux-issues been solved? the shortcuts-feature was proposed eatlier, but that specific implementation has been rejected because it cluttered the ui.

**Comment by AlvaroBrey** [comment]
> However, have the design/ui/ux-issues been solved? the shortcuts-feature was proposed eatlier, but that specific implementation has been rejected because it cluttered the ui.

So after thinking about this a bit, I don't think this is the right place for adding a widget. The share menus are for... well, sharing, which is not what adding a widget is. 

@jancborchardt had the right idea in https://github.com/nextcloud/android/pull/5522 ; I think we should just provide a 1x1 widget that can be configured to open a specific folder, instead. If "pin to homescreen" was to be added somewhere that would be in the overflow menu, but that's overcrowded already. Maybe this can be revisited when we eventually transform that menu into a bottomsheet.

**Comment by AlvaroBrey** [comment]
@jancborchardt what do you think of this? I'd honestly even prefer it in the overflow menu, where it makes more sense conceptually, but it does look a bit overcrowded already.

**Comment by newhinton** [comment]
Should we tag the design team?

**Comment by AlvaroBrey** [comment]
@nextcloud/designers

**Comment by stefan-niedermann** [comment]
> @jancborchardt had the right idea in https://github.com/nextcloud/android/pull/5522 ; I think we should just provide a 1x1 widget that can be configured to open a specific folder, instead.

I can only **strongly** recommend you to **not** hack a shortcut (which it *is*) with a 1x1 widget. We did this mistake once with the [Notes Android app](https://github.com/stefan-niedermann/nextcloud-notes) and suffered by
- badly scaling graphics
- missing long press actions (like going to the apps details)
- missing text
- inconsistent behavior with the other app launchers
- missing possibility to move it to the android bottom bar for the ~5 most important apps
- ...

**Please. Don't misuse widgets as shortcuts.**

---

About this actual PR proposal:

This is not the first place I would looking for shortcuts and while one can argue that it behaves a bit like the Windows context menu (<kbd>*send* to desktop</kbd> for creating a link there = <kbd>*share* to homescreen</kbd> for creating a shortcut there), it honestly feels a bit off.

**I already proposed another possible solution which does effectively prevent cluttering any existing UI elements which I'd like to bring up here again as an alternative: https://github.com/nextcloud/android/issues/6363**

**Comment by newhinton** [comment]
@stefan-niedermann I fully agree what you said about widgets.


I disagree (to some extend) with your proposal. In itself, its a great idea, and should be done, but it does not adress one usecase that you mentioned. There is simply no way to quickly add a shortcut from where you are. You always have to know that the "Add Shortcut" menu exist (which many dont) and even if, you have to go back and find it again. It simply cannot do what this pr would allow the user todo. (That does not mean it has to be merged, but it is a sligthly different but complementary issue to yours)

**Comment by newhinton** [comment]
@AlvaroBrey Any news on this? If you like i can also move it to the overflow as you mentioned.

**Comment by AlvaroBrey** [comment]
> @AlvaroBrey Any news on this? If you like i can also move it to the overflow as you mentioned.

We're going to be changing the overflow dropdown into a bottom sheet soon. Once that's done, that would be the appropriate place for a shortcut functionality, I think, as a bottom sheet can contain more options without losing too much usability.

**Comment by newhinton** [comment]
Is there an ETA for this? I could really use this functionality to be honest. I am willing to do the work for it.

**Comment by AlvaroBrey** [comment]
> Is there an ETA for this? I could really use this functionality to be honest. I am willing to do the work for it.

_should_ be the next thing I work on, barring unexpected urgent things. I can give no ETA other than _Soon™_

**Comment by AlvaroBrey** [comment]
@newhinton The file actions menu redesign has been merged: https://github.com/nextcloud/android/pull/11008

**Comment by newhinton** [comment]
From the screenshots i have seen it looks great! I will take time on the weekend to port this feature.

**Comment by newhinton** [comment]
@AlvaroBrey I have created a new PR since it was a bit easier and i needed to redo the ui-part anyway. Also they are not quite the same from a ux-standpoint.

**Pull request closed without merging**